File: mzcd3-code/code/iCloud/PPRecipes/PPRDataController.m
/***
 * Excerpted from "Core Data in Objective-C, Third Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/mzcd3 for more book information.
***/
//
//  PPRDataController.m
//  PPRecipes
//
//  Created by Marcus S. Zarra on 10/15/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

#import "PPRDataController.h"

#define EVENTS_ARRAY @[UIApplicationDidEnterBackgroundNotification, \
                       UIApplicationWillResignActiveNotification, \
                       UIApplicationWillTerminateNotification]

@interface PPRDataController()

@property (nonatomic, copy) void (^initBlock)(void);

- (void)initializeCoreDataStack;
- (void)initializeDocument;

@end

@implementation PPRDataController

- (instancetype)initWithCompletion:(void(^)(void))block
{
  if (!(self = [super init])) return nil;

  [self setInitBlock:block];

  [self initializeCoreDataStack];

  for (id event in EVENTS_ARRAY) {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(saveContext)
                                                 name:event
                                               object:nil];
  }

  return self;
}

- (void)saveContext
{
  if ([self managedDocument]) {
    NSURL *fileURL = [[self managedDocument] fileURL];
    [[self managedDocument] saveToURL:fileURL
                     forSaveOperation:UIDocumentSaveForCreating
                    completionHandler:^(BOOL success) {
                      //Handle failure
                    }];
    return;
  }

  NSManagedObjectContext *moc = [self managedObjectContext];

  if (!moc) return;
  if (![moc hasChanges]) return;

  NSError *error = nil;
  ZAssert([moc save:&error], @"Error saving MOC: %@\n%@", [error localizedDescription], [error userInfo]);
}


- (void)documentStateChanged:(NSNotification*)notification
{
  switch ([[notification object] documentState]) {
    case UIDocumentStateNormal:
      DLog(@"UIDocumentStateNormal");
      break;
    case UIDocumentStateClosed:
      DLog(@"UIDocumentStateClosed %@", notification);
      break;
    case UIDocumentStateInConflict:
      DLog(@"UIDocumentStateInConflict %@", notification);
      break;
    case UIDocumentStateSavingError:
      DLog(@"UIDocumentStateSavingError %@", notification);
      break;
    case UIDocumentStateEditingDisabled:
      DLog(@"UIDocumentStateEditingDisabled %@", notification);
      break;
    case UIDocumentStateProgressAvailable:
      DLog(@"UIDocumentStateProgressAvailable %@", notification);
      break;
  }
}

- (void)mergePSCChanges:(NSNotification*)notification
{
  NSManagedObjectContext *moc = [self managedObjectContext];
  [moc performBlock:^{
    [moc mergeChangesFromContextDidSaveNotification:notification];
  }];
}

#pragma mark - Core Data stack

- (void)initializeDocument
{
  dispatch_queue_t queue;
  queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);

  dispatch_async(queue, ^{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *storeURL = nil;
    storeURL = [[fileManager URLsForDirectory:NSDocumentDirectory
                                    inDomains:NSUserDomainMask] lastObject];
    storeURL = [storeURL URLByAppendingPathComponent:@"PPRecipes"];

    NSURL *cloudURL = [fileManager URLForUbiquityContainerIdentifier:nil];

    NSMutableDictionary *options = [[NSMutableDictionary alloc] init];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSMigratePersistentStoresAutomaticallyOption];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSInferMappingModelAutomaticallyOption];

    if (cloudURL) {
      cloudURL = [cloudURL URLByAppendingPathComponent:@"PPRecipes"];

      [options setValue:[[NSBundle mainBundle] bundleIdentifier]
                 forKey:NSPersistentStoreUbiquitousContentNameKey];
      [options setValue:cloudURL
                 forKey:NSPersistentStoreUbiquitousContentURLKey];
    }

    UIManagedDocument *document = nil;
    document = [[UIManagedDocument alloc] initWithFileURL:storeURL];
    [document setPersistentStoreOptions:options];

    NSMergePolicy *policy = [[NSMergePolicy alloc] initWithMergeType:
                             NSMergeByPropertyObjectTrumpMergePolicyType];
    [[document managedObjectContext] setMergePolicy:policy];

    void (^completion)(BOOL) = ^(BOOL success) {
      if (!success) {
        ALog(@"Error saving %@\n%@", storeURL, [document debugDescription]);
        return;
      }


      if ([self initBlock]) {
        dispatch_sync(dispatch_get_main_queue(), ^{
          [self initBlock]();
        });
      }
    };

    if ([[NSFileManager defaultManager] fileExistsAtPath:[storeURL path]]) {
      [document openWithCompletionHandler:completion];
      return;
    }

    [document saveToURL:storeURL
       forSaveOperation:UIDocumentSaveForCreating
      completionHandler:completion];
    [self setManagedDocument:document];
  });
}

- (void)initializeCoreDataStack
{
  NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"PPRecipes"
                                            withExtension:@"momd"];
  ZAssert(modelURL, @"Failed to find model URL");

  NSManagedObjectModel *mom = nil;
  mom = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];

  NSPersistentStoreCoordinator *psc = nil;
  psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];

  NSManagedObjectContext *moc = nil;
  NSUInteger type = NSMainQueueConcurrencyType;
  moc = [[NSManagedObjectContext alloc] initWithConcurrencyType:type];
  [moc setPersistentStoreCoordinator:psc];

  [self setManagedObjectContext:moc];

  dispatch_queue_t queue;
  queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
  dispatch_async(queue, ^{
    NSMutableDictionary *options = [[NSMutableDictionary alloc] init];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSMigratePersistentStoresAutomaticallyOption];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSInferMappingModelAutomaticallyOption];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *cloudURL = [fileManager URLForUbiquityContainerIdentifier:nil];
    if (cloudURL) {
      DLog(@"iCloud enabled: %@", cloudURL);
      cloudURL = [cloudURL URLByAppendingPathComponent:@"PPRecipes"];
      [options setValue:[[NSBundle mainBundle] bundleIdentifier]
                 forKey:NSPersistentStoreUbiquitousContentNameKey];
      [options setValue:cloudURL
                 forKey:NSPersistentStoreUbiquitousContentURLKey];
    } else {
      DLog(@"iCloud is not enabled");
    }
    NSURL *sURL = nil;
    sURL = [[fileManager URLsForDirectory:NSDocumentDirectory
                                inDomains:NSUserDomainMask] lastObject];
    sURL = [sURL URLByAppendingPathComponent:@"PPRecipes-iCloud.sqlite"];
    NSError *error = nil;
    NSPersistentStoreCoordinator *coordinator = nil;
    coordinator = [[self managedObjectContext] persistentStoreCoordinator];
    NSPersistentStore *store = nil;




    
    store = [coordinator addPersistentStoreWithType:NSSQLiteStoreType
                                      configuration:nil
                                                URL:sURL
                                            options:options
                                              error:&error];
    if (!store) {
      ALog(@"Error adding persistent store to coordinator %@\n%@",
           [error localizedDescription], [error userInfo]);
      //Present a user facing error
    }
    if ([self initBlock]) {
      dispatch_sync(dispatch_get_main_queue(), ^{
        [self initBlock]();
      });
    }
  });
}

- (void)initializeCoreDataStackV2
{
  NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"PPRecipes"
                                            withExtension:@"momd"];
  ZAssert(modelURL, @"Failed to find model URL");

  NSManagedObjectModel *mom = nil;
  mom = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
  ZAssert(mom, @"Failed to initialize model");

  NSPersistentStoreCoordinator *psc = nil;
  psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];
  ZAssert(psc, @"Failed to initialize persistent store coordinator");

  NSManagedObjectContext *moc = nil;
  moc = [NSManagedObjectContext alloc];
  moc = [moc initWithConcurrencyType:NSMainQueueConcurrencyType];
  [moc setPersistentStoreCoordinator:psc];

  [self setManagedObjectContext:moc];

  dispatch_queue_t queue;
  queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
  dispatch_async(queue, ^{
    NSMutableDictionary *options = [[NSMutableDictionary alloc] init];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSMigratePersistentStoresAutomaticallyOption];
    [options setValue:[NSNumber numberWithBool:YES]
               forKey:NSInferMappingModelAutomaticallyOption];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *docURL = nil;
    docURL = [[fileManager URLsForDirectory:NSDocumentDirectory
                                  inDomains:NSUserDomainMask] lastObject];
    NSURL *storeURL = nil;

    NSError *error = nil;
    NSPersistentStoreCoordinator *coordinator = nil;
    coordinator = [[self managedObjectContext] persistentStoreCoordinator];
    NSPersistentStore *store = nil;

    NSURL *cloudURL = [fileManager URLForUbiquityContainerIdentifier:nil];
    if (!cloudURL) {
      storeURL = [docURL URLByAppendingPathComponent:@"PPRecipes.sqlite"];
      store = [coordinator addPersistentStoreWithType:NSSQLiteStoreType
                                        configuration:nil
                                                  URL:storeURL
                                              options:options
                                                error:&error];
      if (!store) {
        ALog(@"Error adding persistent store to coordinator %@\n%@",
             [error localizedDescription], [error userInfo]);
        //Present a user facing error
        return;
      }
      if ([self initBlock]) {
        dispatch_sync(dispatch_get_main_queue(), ^{
          [self initBlock]();
        });
      }
      return;
    }
    storeURL = [docURL URLByAppendingPathComponent:@"PPRecipes-iCloud.sqlite"];
    NSURL *oldURL = nil;
    oldURL = [docURL URLByAppendingPathComponent:@"PPRecipes.sqlite"];
    if ([fileManager fileExistsAtPath:[oldURL path]]) {
      store = [coordinator addPersistentStoreWithType:NSSQLiteStoreType
                                        configuration:nil
                                                  URL:oldURL
                                              options:options
                                                error:&error];
      if (!store) {
        ALog(@"Error adding OLD persistent store to coordinator %@\n%@",
             [error localizedDescription], [error userInfo]);
        //Present a user facing error
        return;
      }
    }
    cloudURL = [cloudURL URLByAppendingPathComponent:@"PPRecipes"];
    [options setValue:[[NSBundle mainBundle] bundleIdentifier]
               forKey:NSPersistentStoreUbiquitousContentNameKey];
    [options setValue:cloudURL
               forKey:NSPersistentStoreUbiquitousContentURLKey];
    store = [coordinator migratePersistentStore:store
                                          toURL:storeURL
                                        options:options
                                       withType:NSSQLiteStoreType
                                          error:&error];
    if (!store) {
      ALog(@"Error adding OLD persistent store to coordinator %@\n%@",
           [error localizedDescription], [error userInfo]);
      //Present a user facing error
      return;
    }

    ZAssert([fileManager removeItemAtURL:oldURL error:&error],
            @"Failed to remove old persistent store at %@\n%@\n%@",
            oldURL, [error localizedDescription], [error userInfo]);

    if ([self initBlock]) {
      dispatch_sync(dispatch_get_main_queue(), ^{
        [self initBlock]();
      });
    }
  });
}


@end

File: mzswift-code/code/iCloud/PPRecipes/PPRDataController.swift
//
//  PPRDataController.swift
//  PPRecipes
//
//  Created by Marcus S. Zarra on 12/7/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

import Foundation
import CoreData
import UIKit

let RECIPE_TYPES = "ppRecipeTypes"

class PPRDataController : NSObject {
  var mainContext: NSManagedObjectContext?
  var writerContext: NSManagedObjectContext?
  var managedDocument: UIManagedDocument?
  var persistenceInitialized = false
  var initializationComplete: (() -> Void)?

  init(completion: @escaping () -> Void) {
    super.init()
    initializationComplete = completion
    initializeCoreDataStack()
  }

  func initializeCoreDataStackALT() {
    guard let modelURL = Bundle.main.url(forResource: "PPRecipes",
      withExtension: "momd") else {
      fatalError("Failed to locate DataModel.momd in app bundle")
    }
    guard let mom = NSManagedObjectModel(contentsOf: modelURL) else {
      fatalError("Failed to initialize MOM")
    }
    let psc = NSPersistentStoreCoordinator(managedObjectModel: mom)

    var type = NSManagedObjectContextConcurrencyType.privateQueueConcurrencyType
    writerContext = NSManagedObjectContext(concurrencyType: type)
    writerContext?.persistentStoreCoordinator = psc

    type = NSManagedObjectContextConcurrencyType.mainQueueConcurrencyType
    mainContext = NSManagedObjectContext(concurrencyType: type)
    mainContext?.parent = writerContext

    let queue = DispatchQueue.global(qos: .background)
    queue.async {
      let fileManager = FileManager.default
      guard let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
        fatalError("Failed to resolve documents directory")
      }
      let storeURL = documentsURL.appendingPathComponent("PPRecipes.sqlite")

      do {
        try psc.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: storeURL, options: nil)
      } catch {
        fatalError("Failed to initialize PSC: \(error)")
      }
      self.populateTypeEntities()
      self.persistenceInitialized = true
    }
  }

  func initializeCoreDataStack() {
    guard let modelURL = Bundle.main.url(forResource: "PPRecipes", withExtension: "momd") else {
      fatalError("Failed to locate DataModel.momd in app bundle")
    }
    guard let mom = NSManagedObjectModel(contentsOf: modelURL) else {
      fatalError("Failed to initialize MOM")
    }
    let psc = NSPersistentStoreCoordinator(managedObjectModel: mom)

    mainContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
    mainContext?.persistentStoreCoordinator = psc

    let queue = DispatchQueue.global(qos: .background)
    queue.async {
      let fileManager = FileManager.default
      guard let docURL = fileManager.urls(for: .documentDirectory,
        in: .userDomainMask).first else {
        fatalError("Failed to resolve documents directory")
      }
      let storeURL = docURL.appendingPathComponent("PPRecipes.sqlite")
      do {
        try psc.addPersistentStore(ofType: NSSQLiteStoreType,
          configurationName: nil, at: storeURL, options: nil)
      } catch {
        fatalError("Failed to initialize PSC: \(error)")
      }
      self.populateTypeEntities()
      self.persistenceInitialized = true
      if let closure = self.initializationComplete {
        DispatchQueue.main.async {
          closure()
        }
      }
    }
    let center = NotificationCenter.default
    center.addObserver(self, selector: #selector(mergePSCChanges(_:)),
      name: .NSPersistentStoreDidImportUbiquitousContentChanges,
      object: mainContext)
  }

  func initializeCoreDataStack2() {
    guard let modelURL = Bundle.main.url(forResource: "PPRecipes",
                                         withExtension: "momd") else {
      fatalError("Failed to locate DataModel.momd in app bundle")
    }
    guard let mom = NSManagedObjectModel(contentsOf: modelURL) else {
      fatalError("Failed to initialize MOM")
    }
    let psc = NSPersistentStoreCoordinator(managedObjectModel: mom)

    mainContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
    mainContext?.persistentStoreCoordinator = psc

    let queue = DispatchQueue.global(qos: .background)
    queue.async {
      let fileManager = FileManager.default
      guard let docURL = fileManager.urls(for: .documentDirectory,
        in: .userDomainMask).first else {
          fatalError("Failed to resolve documents directory")
      }

      var options = [String:Any]()
      options[NSMigratePersistentStoresAutomaticallyOption] = true
      options[NSInferMappingModelAutomaticallyOption] = true

      let aCloudURL = fileManager.url(forUbiquityContainerIdentifier: nil)

      guard let cloudURL = aCloudURL else {
        let storeURL = docURL.appendingPathComponent("PPRecipes.sqlite")
        do {
          try psc.addPersistentStore(ofType: NSSQLiteStoreType,
            configurationName: nil, at: storeURL, options: options)
        } catch {
          fatalError("Failed to initialize PSC: \(error)")
        }
        if let closure = self.initializationComplete {
          DispatchQueue.main.async {
            closure()
          }
        }
        return
      }

      let filename = "PPRecipes-iCloud.sqlite"
      let storeURL = docURL.appendingPathComponent(filename)
      let oldURL = docURL.appendingPathComponent("PPRecipes.sqlite")
      let oldURLPath = oldURL.path
      var store: NSPersistentStore? = nil
      if fileManager.fileExists(atPath: oldURLPath) {
        do {
          store = try psc.addPersistentStore(ofType: NSSQLiteStoreType,
            configurationName: nil, at: oldURL, options: options)
        } catch {
          fatalError("Error adding store to PSC:\(error)")
        }
      }

      guard let identifier = Bundle.main.bundleIdentifier else {
        fatalError("BundleIdentifier is nil")
      }
      options[NSPersistentStoreUbiquitousContentNameKey] = identifier
      options[NSPersistentStoreUbiquitousContentURLKey] = cloudURL

      do {
        try psc.migratePersistentStore(store!, to: storeURL,
          options: options, withType: NSSQLiteStoreType)
      } catch {
        fatalError("Failed to migrate store: \(error)")
      }

      if let closure = self.initializationComplete {
        DispatchQueue.main.async {
          closure()
        }
      }
    }
  }

  func mergePSCChanges(_ notification: Notification) {
    guard let moc = self.mainContext else {
      fatalError("Unexpected nil MOC")
    }
    moc.perform() {
      moc.mergeChanges(fromContextDidSave: notification)
    }
  }

  func initializeDocument() {
    let queue = DispatchQueue.global(qos: .background)

    queue.async {
      let fileManager = FileManager.default
      guard let documentsURL = fileManager.urls(for: .documentDirectory,
        in: .userDomainMask).first else {
        fatalError("Failed to resolve documents directory")
      }
      let storeURL = documentsURL.appendingPathComponent("PPRecipes")

      var options = [String:Any]()
      options[NSMigratePersistentStoresAutomaticallyOption] = true
      options[NSInferMappingModelAutomaticallyOption] = true

      if let cloudURL = fileManager.url(forUbiquityContainerIdentifier: nil) {
        let url = cloudURL.appendingPathComponent("PPRecipes")
        let identifier = Bundle.main.bundleIdentifier
        options[NSPersistentStoreUbiquitousContentNameKey] = identifier
        options[NSPersistentStoreUbiquitousContentURLKey] = url
      }

      let document = UIManagedDocument(fileURL: storeURL)
      document.persistentStoreOptions = options

      let type = NSMergePolicyType.mergeByPropertyObjectTrumpMergePolicyType
      let policy = NSMergePolicy(merge: type)
      document.managedObjectContext.mergePolicy = policy

      let completion: (Bool) -> Void = { (success:Bool) -> Void in
        if !success {
          print("Failed to open file: \(storeURL)")
          return
        }

        if let closure = self.initializationComplete {
          DispatchQueue.main.async {
            closure()
          }
        }
      }

      self.managedDocument = document

      let path = storeURL.path

      if FileManager.default.fileExists(atPath: path) {
        document.open(completionHandler: completion)
      } else {
        document.save(to: storeURL, for: .forCreating,
          completionHandler: completion)
      }

      let center = NotificationCenter.default
      center.addObserver(self,
        selector: #selector(PPRDataController.documentStateChanged(_:)),
        name: NSNotification.Name.UIDocumentStateChanged, object: document)
    }
  }

  func saveDocument() {
    guard let fileURL = managedDocument?.fileURL else {
      fatalError("document did not have a URL")
    }
    managedDocument?.save(to: fileURL, for: .forOverwriting,
      completionHandler: { (success) in
        //Handle failure
    })
  }

  func documentStateChanged(_ notification: Notification) {
    guard let doc = notification.object as? UIManagedDocument else {
      fatalError("Unexpected object in notification")
    }
    switch doc.documentState {
    case UIDocumentState():
      print("UIDocumentStateNormal: \(notification)")
    case UIDocumentState.closed:
      print("UIDocumentStateClosed: \(notification)")
    case UIDocumentState.inConflict:
      print("UIDocumentStateInConflict: \(notification)")
    case UIDocumentState.savingError:
      print("UIDocumentStateSavingError: \(notification)")
    case UIDocumentState.editingDisabled:
      print("UIDocumentStateDisabled: \(notification)")
    case UIDocumentState.progressAvailable:
      print("UIDocumentStateProgressAvailable: \(notification)")
    default: break
    }
  }

  fileprivate func populateTypeEntities() {
    let pMOC = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
    pMOC.parent = mainContext
    pMOC.performAndWait() {
      let fetch = NSFetchRequest<NSManagedObject>(entityName: "Type")

      do {
        let count = try pMOC.count(for: fetch)
        if count > 0 { return }
      } catch {
        fatalError("Failed to count receipe types: \(error)")
      }

      guard let types = Bundle.main.infoDictionary?[RECIPE_TYPES] as? [String] else {
        fatalError("Failed to find default RecipeTypes in Info.plist")
      }

      for type in types {
        let object = NSEntityDescription.insertNewObject(forEntityName: "Type", into: pMOC)
        object.setValue(type, forKey:"name")
      }

      do {
        try pMOC.save()
      } catch {
        fatalError("Failed to save private child moc: \(error)")
      }
    }
  }

  func saveContext() {
    guard let main = mainContext else {
      fatalError("save called before mainContext is initialized")
    }
    main.performAndWait({
      if !main.hasChanges { return }
      do {
        try main.save()
      } catch {
        fatalError("Failed to save mainContext: \(error)")
      }
    })
    guard let writer = writerContext else {
      return
    }
    writer.perform({
      if !writer.hasChanges { return }
      do {
        try writer.save()
      } catch {
        fatalError("Failed to save writerContext: \(error)")
      }
    })
  }
}
